internal/pkg/pluginengine/change.go

```diff
@@ -169,3 +182,46 @@ func handleResult(smgr statemanager.Manager, change *Change) error {
 	log.Successf("Plugin %s %s done.", change.Tool.Name, change.ActionName)
 	return nil
 }
+
+// FillInputParams fill inputs from state
+func FillInputParams(smgr statemanager.Manager, options map[string]interface{}) error {
+	//traverse options
+	for key, value := range options {
+		log.Debug("Key: ", key, ", Value: ", value)
+		// judge wheter the value is string
+		if inst, ok := value.(string); ok {
+			// judge wheter the format is ${{}}
+			if strings.HasPrefix(inst, REF_PREFIX) && strings.HasSuffix(inst, REF_SUFFIX) {
+				ref := strings.TrimSpace(strings.TrimSuffix(strings.TrimPrefix(inst, REF_PREFIX), REF_SUFFIX))
+				log.Debug("Ref inputs: ", ref)
+				refParam := strings.Split(ref, ".")
+				if len(refParam) <= 3 {
+					return errors.New("ref input format is not correct: " + ref)
+				}
+
+				state := smgr.GetState(genStateKey(refParam))
+				outputs := state.Resource["outputs"]
+				log.Debug("Ref outputs: ", outputs)
+				if outs, ok := outputs.(map[string]interface{}); ok {
+					log.Debug("Ref outs: ", outs)
+					log.Debug("Ref param: ", refParam[3])
+					options[key] = outs[refParam[3]]
+					if value == nil {
+						return errors.New("ref input value is null: " + refParam[3])
+					}
+				}
+			}
+		}
+		// judge wheter the format is map[string]interface{}, if true, then recursion
```

Comment:
I do not understand here. I thought we have already replaced outputs to their values in the previous loop. What is the purpose of this recursion?

Reply:
Because the structure in options may be as follows, not necessarily the first layer can find ${{xxx}}, maybe the second layer, maybe the third layer.

eg.
1. in first layer:
  options:
      doingListId: ${{ default.trello.outputs.doingListId }}
      doneListId: ${{ default.trello.outputs.doneListId }}

2 in second layer:
  options:
    app:
      name: good_name
      doingListId: ${{ default.trello.outputs.doingListId }}
      doneListId: ${{ default.trello.outputs.doneListId }}

Reply:
I do not understand. Could you give a concrete example with a full config.yaml?

Correct me if I'm wrong:

If B depends on A and B uses A's outputs, A will be executed first. When we execute B, we can already get all the outputs in one pass. Even if C depends on B and uses B's outputs, by the time C gets executed, B already finished, so we can fill C's input with one pass.